src/julia_internal.h

```diff
@@ -47,16 +47,26 @@
 #endif
 
 #if jl_has_builtin(__builtin_assume)
-static inline void jl_assume_(int cond)
+#define jl_assume(cond) (__extension__ ({               \
+                __typeof__(cond) cond_ = (cond);        \
+                __builtin_assume(!!(cond_));            \
+                cond;                                   \
+            }))
+#elif defined(_COMPILER_MICROSOFT_) && defined(__cplusplus)
+template<typename T>
+static inline T
+jl_assume(T v)
 {
-    __builtin_assume(cond);
+    __assume(!!v);
+    return v;
 }
+#elif defined(_COMPILER_INTEL_)
 #define jl_assume(cond) (__extension__ ({               \
                 __typeof__(cond) cond_ = (cond);        \
-                jl_assume_(!!(cond_));                  \
-                cond;                                   \
+                __assume(!!(cond_));                    \
```

Comment:
I was confused why GitHub was highlighting trailing white space, but it wasn't. 

Ah s/__assume/__builtin_assume/